lib/forspell/reporter.rb

```diff
@@ -0,0 +1,105 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'pastel'
+require 'logger'
+
+module Forspell
+  class Reporter
+    SUCCESS_CODE = 0
+    ERROR_CODE = 1
+
+    attr_reader :total_errors
+
+    def initialize(logfile:,
+                   verbose:,
+                   format:,
+                   group:)
+
+      FileUtils.touch(logfile) if logfile.is_a?(String)
+      @logger = Logger.new(logfile || STDOUT)
+      @error_logger = Logger.new(STDERR)
+
+      [@logger, @error_logger].each do |logger|
+        logger.formatter = proc do |_severity, _datetime, _progname, msg|
+          "#{msg}\n"
+        end
+      end
+
+      @verbose = verbose
+      @format = group ? 'group' : format
+
+      @pastel = Pastel.new(enabled: $stdout.tty?)
+      @errors = []
+    end
+
+    def file(path)
+      @logger.info "Processing #{path}" if @verbose
+      @files_count ||= 0
+      @files_count += 1
+      @current_file = path
+    end
+
+    def error(word, suggestions)
+      if @format != 'readable'
+        @errors << format(word, suggestions)
+      else
+        @total_errors ||= 0
+        @total_errors += 1
+        @logger.info format(word, suggestions)
+      end
```

Comment:
If you always push the error into `@errors` it would be less `if`s:
```ruby
@errors << [word, suggestions]
output immediately if format == 'readable'
```
(From `@errors`, you can also calculate "total errors count" at the end ;)